### PR TITLE
fix #2615

### LINKED ISSUE #2615: Make Approximation serializeable
If I have a (lengthy) advi optimization, whats the best way to store the result for a future notebook?

```python
with pymc_model:

    approx = pm.fit(
        method='advi', n=50000, random_seed=1, callbacks=cb,
        obj_optimizer=pm.adagrad_window(learning_rate=.5, n_win=100))

    trace = approx.sample(1000)
```

Pickling the trace works,
```python
import pickle

with open('trace.p', 'wb') as f:
    pickle.dump(trace, f)
    
with open('trace.p', 'rb') as f:
    test1 = pickle.load(f)
```

But pickling the approx class fails when I try to load it.

```python
with open('approx.p', 'wb') as f:
    pickle.dump(approx, f)
    
with open('approx.p', 'rb') as f:
    test2 = pickle.load(f)
```
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-31-54dd7d55c9e6> in <module>()
      3 
      4 with open('approx.p', 'rb') as f:
----> 5     test2 = pickle.load(f)

TypeError: __new__() missing 1 required positional argument: 'group'
```

## Versions and main components

* PyMC3 Version: 3.1
* Theano Version: 0.9.0
* Python Version: 3.6
* Operating system: Mac OS X
* How did you install PyMC3: from source

### PR DESCRIPTION
**Pull request by ferrine**
Fixes #2615

### CODE REVIEW

**Comment by ferrine** [comment]
@pstjohn could you please check if this resolves the your issue?

**Comment by pstjohn** [comment]
Does this script work for you?
```python
import pymc3 as pm

with pm.Model() as model:
    pm.HalfNormal('one', shape=(10, 2), total_size=100)
    pm.Normal('two', shape=(10, ))
    pm.Normal('three', shape=(10, 1, 2))

    approx = pm.fit(method='advi', n=5000)
    trace = approx.sample(1000)

import pickle
f = pickle.dumps(approx)
test2 = pickle.loads(f)

test2.sample(1000)
```

When I clone your branch I'm still getting the recursion error on the unpickling line
```
Average Loss = 16.992: 100%|███████████████████████████████████████████████| 5000/5000 [00:00<00:00, 6060.17it/s]
Finished [100%]: Average Loss = 16.981
Traceback (most recent call last):
  File "/Users/pstjohn/Packages/pymc3/pymc3/variational/approximations.py", line 540, in __getattr__
    return getattr(self.groups[0], item)
  File "/Users/pstjohn/Packages/pymc3/pymc3/variational/approximations.py", line 540, in __getattr__
    return getattr(self.groups[0], item)
  File "/Users/pstjohn/Packages/pymc3/pymc3/variational/approximations.py", line 540, in __getattr__
    return getattr(self.groups[0], item)
  [Previous line repeated 996 more times]
RecursionError: maximum recursion depth exceeded while calling a Python object
```

**Comment by ferrine** [comment]
I have the same problem, thanks

File: pymc3/tests/test_variational_inference.py [hunk 1]
```diff
@@ -701,6 +701,18 @@ def test_rowwise_approx(three_var_model, parametric_grouped_approxes):
             pytest.skip('Does not support rowwise grouping')
 
 
+def test_pickle_approx(three_var_approx):
+    import pickle
+    dump = pickle.dumps(three_var_approx)
+    pickle.loads(dump)
```
Comment (pstjohn) [inline pymc3/tests/test_variational_inference.py]: Any reason these don't get run by travis?

**Review by pstjohn — COMMENTED** [review]
No worries, thanks for looking into this!

**Comment by ferrine** [comment]
Seems like my test works fine and SingleGroupApproximation pickles

**Review by pstjohn — APPROVED** [review]
Awesome, can confirm it works on my end too. Thanks a bunch!! That's a super useful feature when you want to store ADVI results either for reproducibility or to run the computation on a different machine.

**Pull request merged**